### PR TITLE
Add support for debug macro information (DW_MACINFO_define/undef)

### PR DESCRIPTION
**Pull request by mgcarrasco**
Implement bidirectional translation of LLVM [DIMacro](https://llvm.org/doxygen/classllvm_1_1DIMacro.html) to SPIR-V [DebugMacroDef ](https://registry.khronos.org/SPIR-V/specs/unified1/DebugInfo.html#DebugMacroDef)and [DebugMacroUndef](https://registry.khronos.org/SPIR-V/specs/unified1/DebugInfo.html#DebugMacroUndef) instructions.

LLVM to SPIR-V:
- Add transDbgMacros() to process all macro definitions and undefs
- Add transDbgMacroDefine() to translate DW_MACINFO_define to DebugMacroDef
- Add transDbgMacroUndef() to translate DW_MACINFO_undef to DebugMacroUndef
- DW_MACINFO_define are processed first, so they are available for DW_MACINFO_undef cases.

SPIR-V to LLVM:
- Add transMacroDef() to translate DebugMacroDef back to DIMacro
- Add transMacroUndef() to translate DebugMacroUndef back to DIMacro
- Add getOrCreateMacroFile() to manage DIMacroFile creation
- Handle edge case where LLVM IR allows undefining macros that were never defined (DebugMacroUndef with DebugInfoNone reference)

### CODE REVIEW

File: test/DebugInfo/DebugMacroDef.ll [hunk 1]
```diff
@@ -0,0 +1,55 @@
+; Test round-trip translation of debug macro information:
+; LLVM IR -> SPIR-V -> LLVM IR
+
+; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-100 %s -o %t.spv
```
Comment (MrSidims) [inline test/DebugInfo/DebugMacroDef.ll]: > Implement bidirectional translation of LLVM [DIMacro](https://llvm.org/doxygen/classllvm_1_1DIMacro.html) to SPIR-V [DebugMacroDef ](https://registry.khronos.org/SPIR-V/specs/unified1/DebugInfo.html#DebugMacroDef)and [DebugMacroUndef](https://registry.khronos.org/SPIR-V/specs/unified1/DebugInfo.html#DebugMacroUndef) instructions.

Note, that the spec links in the PR description are leading to OpenCL.DebugInfo.100 instruction set, while:
a. tests are covering nonsemantic debug info; 
b. code is also build for nonsemantic, see few Literal parameters of the instructions in the referenced OpenCL spec, while in the implementation (and [nonsemantic variant](https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#DebugMacroDef)) there operands are Ids.

I'd suggest to test and implement all 3 debug information types: OpenCL, nonsemantic.100 and nonsemantic.200. The later 2 are identical for Macro handling. OpenCL is a bit different, but not that drastically: see the following lines here and there which transforms literals to constants:
```
    if (isNonSemanticDebugInfo())
      transformToConstant(Ops, {ScopeLineIdx});
```

Comment (mgcarrasco) [reply test/DebugInfo/DebugMacroDef.ll]: thanks, will do.

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 2]
```diff
@@ -1681,3 +1683,75 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgModule(const DIModule *Module) {
...
+}
+
+SPIRVEntry *LLVMToSPIRVDbgTran::transDbgMacroDefine(const DIMacro *Macro,
+                                                    SPIRVString *FileName) {
+  SPIRVWordVec Ops;
+  using namespace SPIRVDebug::Operand::MacroDef;
+  Ops.resize(OperandCount);
+  Ops[SourceIdx] = FileName->getId();
+  Ops[LineIdx] =
+      SPIRVWriter->transValue(getUInt(M, Macro->getLine()), nullptr)->getId();
```
Comment (maarquitos14) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: If I'm looking correctly at the spec, `Line` is a literal number, not an `<id>`.

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 3]
```diff
@@ -1681,3 +1683,75 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgModule(const DIModule *Module) {
...
+  return MacroEntry;
+}
+
+SPIRVEntry *LLVMToSPIRVDbgTran::transDbgMacroUndef(const DIMacro *Macro,
+                                                   SPIRVString *FileName) {
+  SPIRVWordVec Ops;
+  using namespace SPIRVDebug::Operand::MacroUndef;
+  Ops.resize(OperandCount);
+  Ops[SourceIdx] = FileName->getId();
+  Ops[LineIdx] =
```
Comment (maarquitos14) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: Same than above.

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 4]
```diff
@@ -1681,3 +1683,75 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgModule(const DIModule *Module) {
...
+      SPIRVWriter->transValue(getUInt(M, Macro->getLine()), nullptr)->getId();
+
+  SPIRVId MacroDefId = getDebugInfoNoneId();
+
+  // transDbgMacroDefine is processed first so MacroDefMap is already populated
+  // at this point
```
Comment (maarquitos14) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp]:
```suggestion
  // at this point.
```

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp (hunk 2 above)
Comment (mgcarrasco) [reply lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: Thanks! This should be fixed now.

For `isNonSemanticDebugInfo()` we must follow https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#DebugMacroDef where line is an `<id>` otherwise the literal number (https://registry.khronos.org/SPIR-V/specs/unified1/OpenCL.DebugInfo.100.html 4.9 Section).

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp (hunk 3 above)
Comment (mgcarrasco) [reply lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: Thanks, fixed too.

File: test/DebugInfo/DebugMacroDef.ll (hunk 1 above)
Comment (mgcarrasco) [reply test/DebugInfo/DebugMacroDef.ll]: Thanks! This was very helpful. This should be fixed now.

Comment (mgcarrasco) [reply test/DebugInfo/DebugMacroDef.ll]: We are testing all debug information types.

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp (hunk 2 above)
Comment (maarquitos14) [reply lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: Okay, then I think the follow-up is adding SPIRV checks in the tests to check that each debug info type uses the expected format for `Line`.

Comment (mgcarrasco) [reply lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: Thanks, great idea. We are now testing the SPIRV instructions.

**Review by maarquitos14 — APPROVED** [review]

File: lib/SPIRV/libSPIRV/SPIRV.debug.h [hunk 5]
```diff
@@ -971,6 +971,25 @@ enum {
 } // namespace NonSemantic
 } // namespace ImportedEntity
 
+namespace MacroDef {
+enum {
+  SourceIdx      = 0,
+  LineIdx        = 1,
+  NameIdx        = 2,
+  ValueIdx       = 3,
```
Comment (vmaksimo) [inline lib/SPIRV/libSPIRV/SPIRV.debug.h]: `Value` id is optional argument for `DebugMacroDef`. Can we test the case where define is made without the value?

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 6]
```diff
@@ -1681,3 +1683,82 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgModule(const DIModule *Module) {
...
+  Ops.resize(OperandCount);
+  Ops[SourceIdx] = FileName->getId();
+  Ops[LineIdx] = Macro->getLine();
+
+  if (isNonSemanticDebugInfo()){
+    transformToConstant(Ops, {LineIdx});
+  }
+
+  Ops[NameIdx] = BM->getString(Macro->getName().str())->getId();
+  Ops[ValueIdx] = BM->getString(Macro->getValue().str())->getId();
```
Comment (vmaksimo) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: `Value` id is optional argument for `DebugMacroDef`. Can we test the case where define is made without the value?
+also need to handle this in reverse translation.

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp [hunk 7]
```diff
@@ -1407,6 +1407,76 @@ DINode *SPIRVToLLVMDbgTran::transModule(const SPIRVExtInst *DebugInst) {
...
+DIMacro *SPIRVToLLVMDbgTran::transMacroDef(const SPIRVExtInst *DebugInst) {
+  using namespace SPIRVDebug::Operand::MacroDef;
+  const SPIRVWordVec &Ops = DebugInst->getArguments();
+  assert(Ops.size() >= OperandCount && "Invalid number of operands");
+
+  const std::string &SourceFileName = getString(Ops[SourceIdx]);
+  DIFile *File = getDIFile(SourceFileName);
+
+  DIMacroFile *MacroFile =
+      !File ? nullptr : getOrCreateMacroFile(File, DebugInst);
```
Comment (vmaksimo) [inline lib/SPIRV/SPIRVToLLVMDbgTran.cpp:1434]: Just wondering if it makes sense to create "useless" debug info if there's no file to reference?

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp [hunk 8]
```diff
@@ -1407,6 +1407,76 @@ DINode *SPIRVToLLVMDbgTran::transModule(const SPIRVExtInst *DebugInst) {
...
+  // DW_MACINFO_def (i.e., undefining a macro that was never defined).
+  // Example: !DIMacro(type: DW_MACINFO_undef, line: 1, name: "SIZE")
+  //          with no corresponding DW_MACINFO_def for "SIZE"
+  //
+  // When translated to SPIR-V, this becomes:
+  //   DebugMacroUndef <source> <line> DebugInfoNone
+  //
+  // This results in ReferencedMacro being nullptr. We must use a non-empty
+  // fallback name because DIBuilder::createMacro() in LLVM rejects empty
+  // strings.
```
Comment (vmaksimo) [inline lib/SPIRV/SPIRVToLLVMDbgTran.cpp]: Is not this the case where we should drop the debug info? I know the translator can be not the perfect place to deal with the validity of debug info, and such instruction will likely be removed further, but for me it just doesn't make sense to preserve unusable info.. I might missing something here

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp (hunk 7 above)
Comment (mgcarrasco) [reply lib/SPIRV/SPIRVToLLVMDbgTran.cpp:1434]: I think in this particular case it would be incomplete but not useless, just the file would be missing but the other fields will be present.

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp (hunk 8 above)
Comment (mgcarrasco) [reply lib/SPIRV/SPIRVToLLVMDbgTran.cpp]: Thanks. I see your point. The current implementation is not producing meaningful information so it may be better not to emit debug info in this case altogether.

If needed, we could try to handle this more gracefully but in a future PR. I have simplified the code so we don't emit info for this case.

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp (hunk 6 above)
Comment (mgcarrasco) [reply lib/SPIRV/LLVMToSPIRVDbgTran.cpp]: Thanks! Now the code is taking into account that it could be optional.

**Comment by mgcarrasco** [comment]
> The code LGTM, thanks! But it would be nice to have SPIR-V test checks in all the tests, not only in `DebugMacroDef.ll`, sorry for not mentioning this earlier.

@vmaksimo @maarquitos14 thanks for the reviews. All issues have been addressed already, including the SPIRV checks for all tests.

**Review by MrSidims — APPROVED** [review]
